images/hooks/pkg/hooks/create-generic-vmclass/hook.go

```diff
@@ -0,0 +1,174 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package create_generic_vmclass
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"hooks/pkg/settings"
+
+	"github.com/deckhouse/virtualization/api/core"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2"
+
+	"github.com/deckhouse/module-sdk/pkg"
+	"github.com/deckhouse/module-sdk/pkg/registry"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	moduleStateSecretSnapshot = "module-state-secret"
+	vmClassSnapshot           = "vmclass-generic"
+
+	moduleStateSecretName = "module-state"
+	genericVMClassName    = "generic"
+
+	apiVersion = core.GroupName + "/" + v1alpha2.Version
+)
+
+var _ = registry.RegisterFunc(config, Reconcile)
+
+var config = &pkg.HookConfig{
+	OnBeforeHelm: &pkg.OrderedConfig{Order: 5},
+	Kubernetes: []pkg.KubernetesConfig{
+		{
+			Name:       moduleStateSecretSnapshot,
+			APIVersion: "v1",
+			Kind:       "Secret",
+			JqFilter:   `.data`,
+			NameSelector: &pkg.NameSelector{
+				MatchNames: []string{moduleStateSecretName},
+			},
+			NamespaceSelector: &pkg.NamespaceSelector{
+				NameSelector: &pkg.NameSelector{
+					MatchNames: []string{settings.ModuleNamespace},
+				},
+			},
+			ExecuteHookOnSynchronization: ptr.To(false),
+		},
+		{
+			Name:       vmClassSnapshot,
+			APIVersion: apiVersion,
+			Kind:       v1alpha2.VirtualMachineClassKind,
+			JqFilter:   `.metadata.name`,
+			NameSelector: &pkg.NameSelector{
+				MatchNames: []string{genericVMClassName},
+			},
+			LabelSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app":    "virtualization-controller",
+					"module": settings.ModuleName,
+				},
+			},
+			ExecuteHookOnSynchronization: ptr.To(false),
+		},
+	},
+
+	Queue: fmt.Sprintf("modules/%s", settings.ModuleName),
+}
+
+func Reconcile(_ context.Context, input *pkg.HookInput) error {
+	moduleStateSecrets := input.Snapshots.Get(moduleStateSecretSnapshot)
+	vmClasses := input.Snapshots.Get(vmClassSnapshot)
+
+	// nothing to do if generic vmclass already exists
+	if len(vmClasses) > 0 {
+		return nil
+	}
+
+	// if module-state secret exists and contains generic-vmclass-created=true, nothing to do
+	if len(moduleStateSecrets) > 0 {
+		moduleStateData := make(map[string]interface{})
+		if err := moduleStateSecrets[0].UnmarshalTo(&moduleStateData); err != nil {
+			return err
+		}
+
+		if genericCreatedEncoded, exists := moduleStateData["generic-vmclass-created"]; exists {
+			if encodedStr, ok := genericCreatedEncoded.(string); ok {
+				if decodedBytes, err := base64.StdEncoding.DecodeString(encodedStr); err == nil {
+					genericCreated := string(decodedBytes) == "true"
+					if genericCreated {
+						return nil
+					}
+				}
+			}
+		}
```

Comment:
There is a simpler solution: unmarshal into struct with []byte field, so encode/json package will automatically decode base64.

Or unmarshal into corev1.Secret.

```
JqFilter:   `{"metadata": .metadata, "data": .data}`,

...

var moduleStateSecret corev1.Secret

if err := moduleStateSecrets[0].UnmarshalTo[&moduleStateSecret]; err != nil {
 ...
...

// Selected block of 10 lines will be 3 lines only:
if string(moduleStateSecret.Data["generic-vmclass-created"]) == "true" {
   return nil
}
```

Reply:
done